finality-aleph/src/testing/data_store.rs

```diff
@@ -0,0 +1,645 @@
+        .build_and_import_branch_upon(&genesis_hash, MAX_DATA_BRANCH_LEN * 10, false)
+        .await;
+
+    let blocks_branch = blocks[1..2].to_vec();
+    let test_data: TestData = vec![aleph_data_from_blocks(blocks_branch)];
+    test_handler.send_data(test_data.clone());
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not finalized ancestor")
+        .await;
+
+    // After the block gets finalized the message should be let through
+    test_handler.finalize_block(&blocks[0].hash());
+    let message = test_handler
+        .assert_message_out("Did not receive message from Data Store")
+        .await;
+    assert_eq!(message.included_data(), test_data);
+    test_handler.exit();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn correct_messages_go_through_with_late_import() {
+    let (task_handle, mut test_handler) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+    let genesis_hash = test_handler.genesis_hash();
+
+    let blocks = test_handler
+        .build_branch_upon(&genesis_hash, MAX_DATA_BRANCH_LEN * 10)
+        .await;
+    for i in 1..=MAX_DATA_BRANCH_LEN {
+        let blocks_branch = blocks[0..(i as usize)].to_vec();
+        let test_data: TestData = vec![aleph_data_from_blocks(blocks_branch)];
+        test_handler.send_data(test_data.clone());
+    }
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not yet imported blocks")
+        .await;
+
+    test_handler.import_branch(blocks, false).await;
+
+    for _ in 1..=MAX_DATA_BRANCH_LEN {
+        let _message = test_handler
+            .assert_message_out("Did not receive message from Data Store")
+            .await;
+    }
+
+    test_handler.exit();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn message_with_multiple_data_gets_through_when_it_should() {
+    let (task_handle, mut test_handler) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+    let genesis_hash = test_handler.genesis_hash();
+
+    let blocks = test_handler
+        .build_and_import_branch_upon(&genesis_hash, MAX_DATA_BRANCH_LEN * 10, false)
+        .await;
+    let mut test_data = vec![];
+    for i in 1..(MAX_DATA_BRANCH_LEN + 10) {
+        let blocks_branch = blocks[i..(i + 1)].to_vec();
+        test_data.push(aleph_data_from_blocks(blocks_branch));
+    }
+    test_handler.send_data(test_data.clone());
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not finalized ancestor")
+        .await;
+
+    test_handler.finalize_block(&blocks[MAX_DATA_BRANCH_LEN + 7].hash());
+    // This should not be enough yet (ancestor not finalized for some data items)
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not finalized ancestor")
+        .await;
+
+    test_handler.finalize_block(&blocks[MAX_DATA_BRANCH_LEN + 8].hash());
```

Comment:
It is actually not true and `7` is here only because `10-3=7` so a coincidence, but the constants here were indeed confusing. I introduced a single constant now.

Reply:
true, thanks